index.js

```diff
@@ -14,47 +13,86 @@ function createResolver(config, options) {
     resolve: resolve,
   };
 
-  var lastKey;
+  // Keep constants separately
+  var constants = {};
+
+  // Keep requested keys to detect (and disallow) recursive resolution
+  var stack = [];
 
   function resolve(key) {
-    if (typeof key !== 'string') {
-      return;
-    }
 
-    var appliedArgs = slice.call(arguments, 1);
+    if (constants.hasOwnProperty(key)) {
+      return constants[key];
+    }
 
     var definition = config[key];
     // Ignore options that are not defined
     if (!definition) {
       return;
     }
 
-    if (key === lastKey) {
+    if (stack.indexOf(key) >= 0) {
       throw new Error('Recursive resolution denied.');
     }
-    lastKey = key;
 
-    var option = options[key];
-    // Bind the option so it can resolve other options if necessary
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
+    stack.push(key);
+    try {
```

Comment:
This should be abstracted as noted in https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax - try/finally wasn't optimizable until the V8 in node 7.